### PR TITLE
Clarify passing is_causal in sdpa_attention_paged_forward

### PR DESCRIPTION
**Pull request by cyyever**
# What does this PR do?

`is_causal` is not forwarded in `sdpa_attention_paged_forward`. If that is intentional, a comment should be added.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @cyrilvallez!

File: src/transformers/integrations/sdpa_paged.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def sdpa_attention_paged_forward(
         attn_mask=causal_mask,
         dropout_p=dropout,
         scale=scaling,
-        is_causal=False,
+        is_causal=is_causal,
```
Comment (vasqu) [inline src/transformers/integrations/sdpa_paged.py]: I think this might be connected to SDPA's conditions in general, see the non-paged variant https://github.com/huggingface/transformers/blob/6e69b60806ae19f0b45edb96adb40b708172ba7a/src/transformers/integrations/sdpa_attention.py#L76

It is easy to mess up just passing `is_causal` while `False` is guaranteed to work when given the correct attn mask (which we do). In essence, it's an efficiency optimization in cases where we usually don't have a mask.

Comment (vasqu) [reply src/transformers/integrations/sdpa_paged.py]: Imo, should still be aligned somewhat 

CB cc @remi-or @McPatate

Comment (remi-or) [reply src/transformers/integrations/sdpa_paged.py]: `is_causal` is probably not going to work with `sdpa_paged` because we condense the batch of prompts into a size 1 batch, ensuring there is no cross-request contamination with the attention mask, which is not a causal mask. Yet depending on the behavior of `is_causal`, since I can't think of a case where anything above the upper diagonal is True in the ragged attn mask, it might work. But there's maybe some edge cases where it happens, so I would not do this change without a fair amount of testing

Comment (cyyever) [reply src/transformers/integrations/sdpa_paged.py]: @remi-or If `is_causal` is not used, should we remove it?

Comment (remi-or) [reply src/transformers/integrations/sdpa_paged.py]: Good point, I guess we could remove it from the function arguments and leave `is_causal=False` in the call to `torch.nn.functional.scaled_dot_product_attention` . But that might lead to an error if it's used anywhere as an arg. I am not knowledgeable enough about the whole of `transformers` to know if that's the case!

Comment (vasqu) [reply src/transformers/integrations/sdpa_paged.py]: Let's also add a small comment that this is caused by the non-strictly causal attention mask during CB.

Comment (cyyever) [reply src/transformers/integrations/sdpa_paged.py]: @remi-or Absorb it into `kwargs`

Comment (vasqu) [reply src/transformers/integrations/sdpa_paged.py]: > Absorb it into kwargs

We are also doing that for fa2 (non paged) so should be good

Comment (vasqu) [reply src/transformers/integrations/sdpa_paged.py]: > I am not knowledgeable enough about the whole of transformers to know if that's the case!

We are using the module's `is_causal` flag for most things. There is a hidden power feature for base fa2 where you can change causality per kwarg (others rely on the masks to dictate this).

Comment (remi-or) [reply src/transformers/integrations/sdpa_paged.py]: Yes I mean it woud only be a problem if its called as `sdpa_paged(..., True, ...` and not  `sdpa_paged(..., is_causal=True, ...` ! But if we are doing that for other attention then should be good.

Comment (vasqu) [reply src/transformers/integrations/sdpa_paged.py]: Ah no, we never promoted that as positional arg afaik

Comment (cyyever) [reply src/transformers/integrations/sdpa_paged.py]: @remi-or @vasqu  According to the PyTorch documentation, it's not possible to specify `is_causal=True` if `attn_mask` is not None. We only need to be sure that there is not `causal` usage when `attn_mask` is None. But given that it is a general function, it depends on the current usage. However, after searching all usages, it is False.
```
grep 'attention_interface(' -r src/  -C 7 | grep is_causal -C 7 
```

Comment (vasqu) [reply src/transformers/integrations/sdpa_paged.py]: We only enable `is_causal` when the conditions are met, see https://github.com/huggingface/transformers/pull/40838#discussion_r2343789718

They are known but only seldom do we encounter this as no mask (ie no padding) is rare in itself.

Comment (vasqu) [reply src/transformers/integrations/sdpa_paged.py]: Ie see `getattr(module, "is_causal", True)` - we ignore the kwarg there (yes it's messy :sweat_smile:) and check if we can use the `is_causal` kwarg of SDPA:
- No mask
- No decoding step (needs full attn)
- Is marked via the flag as causal model

Comment (vasqu) [reply src/transformers/integrations/sdpa_paged.py]: I wouldn't as per https://github.com/huggingface/transformers/pull/40838#discussion_r2344357560

CB has a hard reliance on the mask (even if it was strictly causal). Just to be sure, @remi-or CB always passes an attention mask?

Comment (cyyever) [reply src/transformers/integrations/sdpa_paged.py]: > Ie see `getattr(module, "is_causal", True)` - we ignore the kwarg there (yes it's messy 😅) and check if we can use the `is_causal` kwarg of SDPA:
> 
> * No mask
> * No decoding step (needs full attn)
> * Is marked via the flag as causal model

Just copy the condition from SPDA to paged SPDA? Then I wonder what is the difference of these two? They all forward to torch SDPA.. Is it possible to merge them into one?

File: src/transformers/integrations/sdpa_paged.py [hunk 2]
```diff
@@ -20,10 +20,9 @@ def sdpa_attention_paged_forward(
     query: torch.Tensor,
     key: torch.Tensor,
     value: torch.Tensor,
-    attention_mask: Optional[torch.Tensor],
+    attention_mask: Optional[Union[torch.Tensor, dict]],
```
Comment (Cyrilvallez) [inline src/transformers/integrations/sdpa_paged.py]: I'm surprised we need code to select from a dict here, usually when entering attention_interface it can only be a Tensor or None
cc @remi-or you added it a few days ago?

File: src/transformers/integrations/sdpa_paged.py [hunk 3]
```diff
@@ -58,6 +57,8 @@ def sdpa_attention_paged_forward(
         attn_mask=causal_mask,
         dropout_p=dropout,
         scale=scaling,
+        # It is assume that either the caller specifies `attention_mask` or in otherwise non-causal context.
+        # See discussions in https://github.com/huggingface/transformers/pull/40838
```
Comment (Cyrilvallez) [inline src/transformers/integrations/sdpa_paged.py]: Not entirely sure this comment is needed, may be a bit misleading IMO - paged implementation uses packed sequence format, so in general it can never be causal in the sdpa sense (except maybe if only 1 sequence is passed, but should be extremely rare for CB use cases)

File: src/transformers/integrations/sdpa_paged.py (hunk 2 above)
Comment (remi-or) [reply src/transformers/integrations/sdpa_paged.py]: I would expect it to be the case but I have not checked across all modeling code. There is something to catch the case where the attention mask is a dict so it would make sense to change the typing, or to remove the `isinstance(attention_mask, dict): ...` branch altogether. Perhaps the second is out of scope for this PR?

File: src/transformers/integrations/sdpa_paged.py (hunk 3 above)
Comment (cyyever) [reply src/transformers/integrations/sdpa_paged.py]: Fixed

File: src/transformers/integrations/sdpa_paged.py (hunk 1 above)
Comment (remi-or) [reply src/transformers/integrations/sdpa_paged.py]: >Just to be sure, @remi-or CB always passes an attention mask?

For SDPA paged, yes. CB only skips the attention mask for `flash_paged`

File: src/transformers/integrations/sdpa_paged.py (hunk 2 above)
Comment (Cyrilvallez) [reply src/transformers/integrations/sdpa_paged.py]: Alright, that's what I thought! Then indeed we should not update this typing (only a Tensor or None is allowed) - I do think we can do it in this PR though, but then it becomes the main goal and the title should be updated

Comment (cyyever) [reply src/transformers/integrations/sdpa_paged.py]: Can do it in another PR

**Review by vasqu — APPROVED** [review]
Let's update the title, as far as I see the typing for the attn mask will be improved in a different (followup) PR

**Comment by vasqu** [comment]
Thank you @cyyever :hugs:
